Hilarity ensued.
```
 t10938.scala:4: error: object xxx is not a member of package scala
+did you mean xml?
   def f(): scala.xxx.XXX[_, _] = ???
                  ^
```

--- a/src/testkit/scala/tools/testkit/BytecodeTesting.scala
+++ b/src/testkit/scala/tools/testkit/BytecodeTesting.scala
@@ -181,6 +181,9 @@ object BytecodeTesting {
   }
 
   def makeSourceFile(code: String, filename: String): BatchSourceFile = new BatchSourceFile(filename, code)
+  private var fileCount = 0
+  private def fileN = { val n = fileCount; fileCount += 1; if (n == 0) "" else n.toString }
+  def SourceFile(code: String*): List[BatchSourceFile] = code.map(makeSourceFile(_, s"UnitTestSource$fileN")).toList
 
   def getGeneratedClassfiles(outDir: AbstractFile): List[(String, Array[Byte])] = {
     def files(dir: AbstractFile): List[(String, Array[Byte])] = {

--- a/src/testkit/scala/tools/testkit/XMLTesting.scala
+++ b/src/testkit/scala/tools/testkit/XMLTesting.scala
@@ -1,0 +1,66 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.tools.testkit
+
+object XMLTesting {
+  object xml {
+    val code = """
+import collection.{immutable, mutable}, mutable.ArrayBuffer
+
+package scala.xml {
+  trait MetaData
+    //def key: String
+    //def value: Seq[Node]
+    //def next: MetaData
+  trait NamespaceBinding
+  object TopScope extends NamespaceBinding
+  object Null extends MetaData
+  abstract class Node extends immutable.Seq[Node] {
+    def label: String
+    def child: Seq[Node] = Nil
+    override def toString = label + child.mkString
+
+    def iterator: Iterator[Node] = ??? // implements `def iterator: Iterator[A]`
+    // Members declared in scala.collection.SeqOps
+    def apply(i: Int): Node = ??? // implements `def apply(i: Int): A`
+    def length: Int = ???
+  }
+  class Elem(prefix: String, val label: String, attributes1: MetaData, scope: NamespaceBinding, minimizeEmpty: Boolean, override val child: Node*) extends Node
+  class NodeBuffer extends Seq[Node] {
+    val nodes = ArrayBuffer.empty[Node]
+    def &+(o: Any): this.type =
+      o match {
+        case n: Node => nodes.addOne(n); this
+        case _ => throw new MatchError(o)
+      }
+    // Members declared in scala.collection.IterableOnce
+    def iterator: Iterator[scala.xml.Node] = nodes.iterator
+    // Members declared in scala.collection.SeqOps
+    def apply(i: Int): scala.xml.Node = nodes(i)
+    def length: Int = nodes.length
+  }
+  case class Text(text: String) extends Node {
+    def label = text
+  }
+  case class Atom(t: Text) extends Node {
+    def label = t.text
+  }
+  trait Attribute extends MetaData
+  class PrefixedAttribute(pre: String, key: String, value: Seq[Node], next1: MetaData) extends Attribute
+  class UnprefixedAttribute(key: String, value: Seq[Node], next1: MetaData) extends Attribute {
+    def this(key: String, value: String, next1: MetaData) = this(key, Text(value), next1)
+  }
+}
+"""
+  }
+}

--- a/test/junit/scala/tools/nsc/parser/ParserTest.scala
+++ b/test/junit/scala/tools/nsc/parser/ParserTest.scala
@@ -8,7 +8,7 @@ import org.junit.Assert._
 import scala.tools.testkit.BytecodeTesting
 
 @RunWith(classOf[JUnit4])
-class ParserTest extends BytecodeTesting{
+class ParserTest extends BytecodeTesting {
   override def compilerArgs: String = "-Ystop-after:parser -Yvalidate-pos:parser -Yrangepos"
   @Test
   def crlfRangePositionXml_t10321(): Unit = {

--- a/test/junit/scala/tools/nsc/typechecker/XMLTest.scala
+++ b/test/junit/scala/tools/nsc/typechecker/XMLTest.scala
@@ -1,0 +1,43 @@
+
+package scala.tools.nsc.typechecker
+
+import org.junit.Test
+import org.junit.Assert._
+
+import scala.tools.nsc.reporters.StoreReporter
+import scala.tools.testkit.{BytecodeTesting, XMLTesting}, BytecodeTesting.SourceFile, XMLTesting.xml
+
+class XMLTest extends BytecodeTesting {
+  override def compilerArgs: String = "-Ystop-after:refchecks -Wnonunit-statement"
+
+  @Test def `building elements does not warn nonunit-statement`: Unit = {
+    val code =
+      """
+        |object Test {
+        |  val xml = <xml><elem></elem></xml>
+        |}
+      """.stripMargin
+    val reporter = compiler.global.reporter.asInstanceOf[StoreReporter]
+    val run = compiler.newRun()
+    run.compileSources(SourceFile(xml.code, code))
+    assertFalse(reporter.hasWarnings)
+    assertFalse(reporter.hasErrors)
+  }
+}
+class XMLLessTest extends BytecodeTesting {
+  override def compilerArgs: String = "-Ystop-after:refchecks"
+
+  @Test def `detect XML lib is absent`: Unit = {
+    val code =
+      """
+        |object Test {
+        |  val xml = <xml><elem></elem></xml>
+        |}
+      """.stripMargin
+    val reporter = compiler.global.reporter.asInstanceOf[StoreReporter]
+    val run = compiler.newRun()
+    run.compileSources(SourceFile(code))
+    assertTrue(reporter.hasErrors)
+    assertTrue(reporter.infos.forall(_.msg.contains("scala.xml")))
+  }
+}

